util/fmtutil/format.go

```diff
@@ -133,25 +133,21 @@ func makeTimeseries(wr *prompb.WriteRequest, labels map[string]string, m *dto.Me
 		toTimeseries(wr, labels, timestamp, float64(m.GetSummary().GetSampleCount()))
 
 	case m.Histogram != nil:
-		metricName := labels[model.MetricNameLabel]
-		// Preserve metric name order with first bucket suffix timeseries then sum suffix timeserie and finally count suffix timeserie
-		// Add Histogram bucket timeseries
-		bucketLabels := make(map[string]string, len(labels)+1)
-		for key, value := range labels {
-			bucketLabels[key] = value
+		h := m.GetHistogram()
+
+		// TODO(mem): using a Schema value as a signal that this is a native histogram.
+		if h.Schema != nil {
+			addNativeHistogramToWriteRequest(wr, h, labels, timestamp)
 		}
-		for _, b := range m.GetHistogram().Bucket {
-			bucketLabels[model.MetricNameLabel] = metricName + bucketStr
-			bucketLabels[model.BucketLabel] = fmt.Sprint(b.GetUpperBound())
-			toTimeseries(wr, bucketLabels, timestamp, float64(b.GetCumulativeCount()))
+
+		// TODO(mem): should this be else? Is a classic histogram without buckets even possible?
```

Comment:
That makes sense, I think.

For a classic histogram, if I don't have buckets, I'm left with the sample count and the sample sum. Looking at the Go code, I can see that the count and the sum will be populated regardless of whether the histogram is native or classic.

With the Go client, if I _try_ to create a histogram without buckets (by specifying an empty slice), I still get the default buckets. The only way (with that current code) that I can find to make it respect the empty bucket list is to make it a native histogram (specify the native histogram bucket factor).

So now the part where I'm confused: I think you are right, a classic histogram without buckets seems to be possible (just by looking at the code), and it will have only the count and sum (and possibly the +inf bucket, by definition?). This does not seem to match the implementation.

Reply:
And just to clarify what I mean by the last paragraph: the code I looked at seems to be fine with an empty bucket list. I cannot spot any assumption that the bucket list is populated. So I _think_ that if the bucket list was in fact empty, the code would work.

Reply:
> Looking at the Go code, I can see that the count and the sum will be populated regardless of whether the histogram is native or classic.

But in case of a native histogram, the count and sum is part of the complex histogram sample, while in the case of the classic histogram, it will spawn to new series (with new names due to "magic" suffixing).

> With the Go client, if I try to create a histogram without buckets (by specifying an empty slice), I still get the default buckets.

IIRC correctly, the intention here was to make users use a Summary without quantiles instead of a Histogram without buckets. But creative users will come up with ways of creating a bucketless classic histogram, I'm sure…
I guess it _should_ have the +Inf bucket then, but if it doesn't, Prometheus will probably tolerate it.

OpenMetrics will have ruled those edge cases more strictly, but I think our code should still be tolerant if a histogram comes in that doesn't comply strictly with the standard, but we can still make sense of it.